### PR TITLE
.Net: Added IChatHistoryReducer for ChatCompletion with AzureOpenAI and OpenAI

### PR DESCRIPTION
**Pull request by sandrohanea**
### Motivation and Context

When using Auto-Invocation for Tools with OpenAI and AzureOpenAI, there is no way to intercept and reduce the chat history after tools are being invoked.

This is needed as some tools might return long responses and we might want to execute different prompt compression / optimization based on the tool result.

Note: `AutoFunctionInvocationFilters` cannot be used as they are invoked only before the result of the tools is added to the chat history.

The alternative to this approach would be to intercept the http call to Open AI and alter the request directly but that will cause multiple allocation, serializations and deserializations and won't scale to other Connectors.

### Description

This change is extracting `IChatHistoryReducer` from `Agents.Core` project and moving it to SK.Abstrations.AI.ChatCompletions.

`IChatHistoryReducer` was renamed to `IAgentChatHistoryReducer` in order to avoid confusion as it requires 2 more methods: `Equals` and `GetHashCode` that are used to compute the agent channel key. It now inherits `IChatHistoryReducer`.

Added tests to ensure that a *DI registered reducer* is used in the ChatCompletionService of both OpenAI and AzureOpenAI connectors.

Important notice: the reducer is not mutating the `ChatHistory`, as it might be that new tool invocation or chat turns will require different reducing logic. (e.g. context-aware summarization) 

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    /// <remarks>
+    /// If there is no <see cref="IChatHistoryReducer"/> registered in the <see cref="Kernel"/>, the original chat history will be returned.
+    /// If the <see cref="IChatHistoryReducer"/> returns null, the original chat history will be returned.
+    /// Note: This is not mutating the original chat history.
+    /// </remarks>
+    /// <param name="chatHistory">The current chat history, including system messages, user messages, assistant messages and tool invocations.</param>
+    /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to cancel this operation.</param>
+    /// <returns>The reduced chat history if the kernel had one reducer configured, or the same <see cref="ChatHistory"/> if it cannot be reduced.</returns>
+    public static async Task<IEnumerable<ChatMessageContent>> TryReduceAsync(this ChatHistory chatHistory, Kernel? kernel, CancellationToken cancellationToken)
```
Comment (crickman) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: Nit - "Try*" methods generally return bool and have an 'out' parameter.  I'm not oppposed to the name change, but do wonder if it doesn't break with convention.

Comment (sandrohanea) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: I agree, named it initially when it was returning bool, but now it makes more sense to be called `ReduceAsync`

File: dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    /// <remarks>
+    /// If there is no <see cref="IChatHistoryReducer"/> registered in the <see cref="Kernel"/>, the original chat history will be returned.
+    /// If the <see cref="IChatHistoryReducer"/> returns null, the original chat history will be returned.
+    /// Note: This is not mutating the original chat history.
+    /// </remarks>
+    /// <param name="chatHistory">The current chat history, including system messages, user messages, assistant messages and tool invocations.</param>
+    /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to cancel this operation.</param>
+    /// <returns>The reduced chat history if the kernel had one reducer configured, or the same <see cref="ChatHistory"/> if it cannot be reduced.</returns>
+    public static async Task<IEnumerable<ChatMessageContent>> ReduceAsync(this ChatHistory chatHistory, Kernel? kernel, CancellationToken cancellationToken)
```
Comment (RogerBarreto) [inline dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: Reduce doesn't need to be asynchronous as it is not using any asynchronous method on its implementation.

```suggestion
    public static IEnumerable<ChatMessageContent> Reducec(this ChatHistory chatHistory, Kernel? kernel, CancellationToken cancellationToken)
```

Comment (sandrohanea) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: The reducer can be Async, so in that case, we need to await the Reduce call:
image

Comment (RogerBarreto) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: This is the same method (recursive), if it's not async, you don't need to `await`.

Comment (sandrohanea) [reply dotnet/src/SemanticKernel.Abstractions/AI/ChatCompletion/ChatHistoryReducerExtensions.cs:28]: No, it's not the same method (just have the same name, but different arguments).

The caller is an extension method on ChatHistory in:[ChatHistoryReducerExtensions.cs](https://github.com/microsoft/semantic-kernel/pull/8894/files/14db8012ba65b668691a62ac5902a5ff4b711f27#diff-8b4255c1d84792dc2a20b999b1967b25bf0ce8bb99f57885fc9bf9ea5bfc76e7)

The callee is an implementation of the method described by interface: [IChatHistoryReducer](https://github.com/microsoft/semantic-kernel/pull/8894/files/14db8012ba65b668691a62ac5902a5ff4b711f27#diff-87234ff20d807c70c611fc61acc1bac9bc65fc97c5321175939773988c843409R23) which is retrieved from the DI container in embedded in the Kernel.

**Comment by markwallace-microsoft** [comment]
Hi @sandrohanea thanks for creating this PR. The team met to discuss and a few things came up

1. We currently mutate `ChatHistory` when automatic function calling is enabled. A function calling loop can be long and have many round trips to the model and truncating the `ChatHistory` could have undesired side effects. In general the consensus was that we don't want to chat history reducer to be called by the chat completion service.
2. The code shows how to use the reducer with Dependency Injection but we also need to handle non DI cases.

The consensus was that developers have control over the `ChatHistory` that they send to the chat completion service so they can reduce it as necessary and also decide which messages from the response get added back into the `ChatHistory`. We're going to work on some samples (during the current sprint) for this we will share with you once we have them.

**Comment by sandrohanea** [comment]
Closing this PR as it's not currently aligning with SK plan for abstractions.

**Comment by qihangnet** [comment]
@markwallace-microsoft Can the Reducer implementation in #7570 be moved to Core? I believe it is useful not just for the Agent.

**Pull request closed without merging**